### ISSUE TITLE
Type mismatches in interface implementations are incorrectly allowed

### ISSUE THREAD

**Issue by bjitivo** [HaxeFoundation/haxe#4378]
Example code:

```
class Data
{
    public var value : String;

    public function new()
    {
        value = "Hello, world!";
    }
}

interface IBase
{
    function updateData(data : Dynamic) : Void;
}

class Base implements IBase
{
    public function updateData(inData : Data) : Void
    {
        trace("inData is: " + inData.value);
    }
}

interface ISub extends IBase
{
}

class Foo extends Base implements ISub
{
    public static function main()
    {
        var ib : IBase = new Foo();

        ib.updateData(7);
    }

    public function new()
    {
    }
}
```

The above is, I contend, invalid code, as the Base class implements this interface function signature:

 function updateData(data : Dynamic) : Void;

via this function:

public function updateData(inData : Data) : Void

The compiler accepts this, but this allows the call as seen in main():

ib.updateData(7);

which causes the number 7 to be passed into Base.updateData instead of an instance of the Data class as expected by this function.

This appears to be an issue going back to at least haxe 3.1.3; here are the results of compiling and running the above:

haxe 3.1.3 compiler, hxcpp back-end: Segmentation fault (core dumped)
haxe 3.1.3 compiler, neko back-end: Uncaught exception - Segmentation fault
haxe 3.2 compiler, hxcpp back-end: hxcpp produces bad C++ code that won't compile
haxe 3.2 compiler, neko back-end: Uncaught exception - Segmentation fault

**Comment by ncannasse** [comment]
It surely should not produce bad C++ code, but on Neko it works as expected:
```
Called from ? line 1
Called from Main.hx line 34
Called from Main.hx line 20
Uncaught exception - Invalid field access : value
```

Which is normal since you're trying to access `.value` field of `7`

**Comment by Simn** [comment]
IMO this violates variance.

**Comment by ncannasse** [comment]
@Simn yes that's what I was looking into atm. I was first concerned about the segfault report on Neko, which I can't reproduce.
There's indeed a variance issue with Dynamic since allowing "only" Data is a restriction wrt the interface specification.

**Comment by waneck** [comment]
If he's compiled himself neko and is using jit (32-bit Linux), there is a
known issue on which it segfaults. There is even a long standing PR I
opened which solves this on the repo.  Don't remember why it hasn't been
merged yet

**Comment by Simn** [comment]
What should we do with this? This is clearly invalid, but Dynamic being Dynamic I don't know if we could give an error here.

**Comment by ncannasse** [comment]
Is it normal that Typeload.valid_redefinition uses Type.unify now? I'm pretty sure I wrote it first using type_eq

**Comment by ncannasse** [comment]
anyway, we could disallow valid (not-Dynamic) (Dynamic) since this is clearly unsound.

**Comment by Simn** [comment]
I think this was changed to unify for Haxe 3 when we allowed contravariance for function arguments. The problem with Dynamic here is that it it unifies in both directions, while technically it's a more general type and should only unify accordingly.

I'm not sure about disallowing this though. We usually don't "care" about Dynamic with regards to variance. How is the redefinition case different from type parameter variance, e.g. here?

```haxe
class Main  {
	static function main() {
		var a = [1];
		var b:Array<Dynamic> = a;
		b.push("foo");
	}
}
```
